lab-bundle-spec.md

```diff
@@ -158,41 +163,63 @@ environment_resources:
------------------------------ | -------- | ------- | --------------------------------------
-dm_template.script            |          | path    | Relative path to a Deployment Manager directory tree.
-dm_template.custom_properties |          | array   | Array of key/value pairs.
-fleet                         |          | enum*   | Specify a Qwiklabs fleet to pull the project from.
+attribute                        | required | type    | notes
+---------------------------------| -------- | ------- | --------------------------------------
+startup_script.type              |          | string  | The type of startup script. Only `deployment_manager` is supported.
+startup_script.path              |          | path    | Relative path to a directory tree with the script contents.
+startup_script.custom_properties |          | array   | Array of pairs. See below for details.
+cleanup_script.type              |          | string  | The type of cleanup script. Only `deployment_manager` is supported.
+cleanup_script.path              |          | path    | Relative path to a directory tree with the script contents.
+cleanup_script.custom_properties |          | array   | Array of pairs. See below for details.
+ssh_key_user                     |          | string  | If this project should use a user's SSH key, the id of that user.
 
 ```yml
 - type: gcp_project
   id: secondary_project
-  fleet: gcpfree
-  dm_template:
-    script: deployment_manager
+  variant: gcpfree
+  startup_script:
+    type: deployment_manager
+    path: dm_startup.zip
     custom_properties:
       - key: userNameWindows
         value: student
+  cleanup_script:
+    type: deployment_manager
+    path: dm_cleanup.zip
+    custom_properties:
+      - key: primary_project_zone
+        value: my_primary_project.zone
 ```
 
-> **NOTE:** Not all GCP fleet names are supported.
->
-> The existing concept of Qwiklabs' Fleets does not have a single analog in
-> content bundles. Notice that some fleet types map to resource types (e.g.
-> `gsuite_multi_tenant` fleet is now the `gsuite-domain` resource type).
-> However other fleets are "variants" of the same resource type (e.g.
-> "free", "ASL", and "standard" GCP projects). Therefore, not all fleet names that
-> an experienced Qwiklabs author may be familiar with, are allowed in the `fleet`
-> field.
+> **NOTE:** The existing concept of Qwiklabs' Fleets does not have a single
+> analog in content bundles.
 >
-> Presently, authors are allowed to specify one of the following fleet types for
-> a gcp_project:
-> - gcpd [default]
-> - gcpfree
-> - gcpasl
->
-> Future versions of the Content Bundle spec may use different terminology for
-> resource_type variations to avoid conflation.
+> Some fleet types map to resource types (e.g. `gsuite_multi_tenant` fleet is
+> now the `gsuite-domain` resource type), while other fleets are allowed as
+> "variants" of `gcp_project` (see below).
+
+###### Variants
```

Comment:
I think variant is a forward-compatible abstraction for fleets but I no longer think it's the complete picture.

I think a project is defined as:
* billing account
* quotas
* APIs

I think variant properly encapsulates all valid (billing_account, APIs) pairs. I'm on the fence on splitting those right now. It would open the door to help reduce our API enablement volume on certain APIs (like ML) but Navy is nowhere near practically supporting arbitrary lists of APIs.

I foresee "quotas" being defined as `cpu_count`, `gpu_count`, `tpu_count`, and `egress_volume`. Given that we don't support customizing any of those right now and adding those fields are all compatible with the current v2 spec, I'd recommend leaving them out entirely until we're ready to seriously support them.

With all of that said, what do you think about replacing variant with:

| variant | billing_account | cpu_count |
| - | - | - |
| gcpd | _default_ | _default_ |
| gcpfree | none | _default |
| gcpondemand | _default_ | 9 |
| gcpedu | edu | _default_ |

Reply:
So to clarify, you're saying there would be no `fleet` or `variant` property. Instead, we would add two properties to the `gcp_project` resource: `billing_account` and `cpu_count`. Is that correct?

If so, I think that has some merit. My thoughts are:
- Decoupling the `billing_account` and `cpu_count` could reduce combinatorial explosion as we unlock more values for these. On the flip side, it prevents us from silently changing these params (e.g. uniting or splitting billing accounts, changing gcpondemand to use 12 cpus, etc). I don't know how likely either of these scenarios (adding values or silently changing them) is though.
- I liked the idea that `variant` would be an attribute that all resource types could have. That goes away in the new proposal.
- I don't like adding more fields that have a seemingly-random set of allowed values.
- I do like that this has obvious extensions when we do allow customizing gpus, tpus, egress volume.

So, tl;dr: I see pros and cons and don't have a strong opinion either way. Would be very curious to hear from others (in particular @jgan and @dcbrandao)

Reply:
i believe we (josh, greg, and i) had a conversation about this sort of thing when we were working on the structure of navy, and we decided that we'd be okay with having variants(/subtypes) that corresponded to each of the combinations of those things that we wished to support

Reply:
I think modern Navy subtypes (which are the same thing as variants) encapsulate supported (billing_account, apis, cpu_count). Going forward, I'm concerned we're going to need to separate out `cpu_count` to help with abuse checking and separate out `apis` to appease certain, problematic GCP services.

Sooo...I think we should split `variant` into a few different fields to support other changes we'd like to eventually make in runtime.

Reply:
because (type, subtype) is so fundamental to navy, what you're suggesting sounds like a major overhaul of the navy backend, unless what you're really intending is we retain the current type identifier but have some mapping from (billing_account, apis, cpu_count) into more subtypes.

that being said, with respect to the content bundle spec, i think listing out the used apis is fine. this is something the author has complete control over, and we can use it immediately for other purposes.
cpu_count seems problematic. we don't currently give authors enough tools to determine what the cpu count should be, i haven't been convinced that cpu count is the proper metric we should be using for abuse checking, and i'm not sure that authors should necessarily need to care about abuse checking
are you thinking of billing_account as some enum where we interpret what the value is supposed to mean, or as some gcp identifier?

Reply:
Derek and I chatted and I think we agreed on:

* `billing_account` is an enum of (none, edu, asl, or default)
* `cpu_count` is an optional integer that defaults to the billing account default value. It's generally 6. If a creator wants `gcpondemand`, set this field to 9.
* `apis` is an optional list of APIs that must be enabled in the project. It defaults to the billing account default value which is ~85 APIs for gcpd. For now, we don't have a way to technically do anything with this list of APIs but we'd like to in the future.

@dcbrandao please double check this.

Reply:
i'm not sure we agreed on cpu_count, i'm still not sold on this manner of specifying abuse checking

Reply:
`billing_account` without `cpu_count` can't distinguish `gcpd` from `gcpondemand`. Also, here's the design doc I was talking about that isn't doing too well in review :) https://docs.google.com/document/d/1HVYxroXhWkaCaOGpg7vSS_Kg7g2LkoRWdcyy54gdFbQ/edit#

Reply:
@dcbrandao not sure how you would specify abuse checking then? Seems like the 2 choices are by amount of resources used (which correlates to project charge) or by checking the students behavior with those resources. Even if the student was performing the lab steps as intended we wouldn't want them to over use resources

Reply:
well, i have some questions about this in general:
 - do we want authors involved in the abuse enforcement (i.e., by specifying what should be considered abuse on the lab)?
 - if so, is cpu count the right metric to give to the lab author wrt abuse enforcement, or is there some other metric or combination of metrics? 
 - should abuse enforcement happen per-resource (which is what cpu count on a given gcp project would be) or across the entire lab?

Reply:
I think there are three options for "should authors be involved in abuse enforcement":
1. Yes, they should have complete control and specify numerical thresholds for all abuse metrics
2. Yes, we should give them a handful of variants that they can pick between that bundle up abuse metrics (status quo)
3. No, runtime should adaptively identify the expected resource usage and tune abuse detection accordingly

I would like to be moving towards (3) in the long-term, but for now I prefer (2). Especially because we can bundle other considerations into the variants to help the authors decide / nudge them in the right direction. Note that if go with (1), the author has no incentive to pick values other than the max for everything.

Reply:
Fascinating! I like (2) for now because it strikes a good balance between effort and reward but I'd like to move towards (1) over time for greater flexibility and letting people lock down labs that would otherwise be too dangerous to use.

As for your point on authors not having the right incentives, I agree as of today but I'm hoping that is not the case long term (cue my rickety design doc on lab locking).

Reply:
We can definitely discuss which direction we want to move, but it sounds like we're agreed that for now we should stick with (2), so I think this issue is resolved.